### ISSUE TITLE
Env vars are compiled into app

### ISSUE THREAD

**Issue by mars** [mars/create-react-app-buildpack#7]
_Updated Oct 16, 2016 with the final solution_
## Problem

When using the [`REACT_APP_*` or `NODE_*` environment vars](https://github.com/mars/create-react-app-buildpack#environment-variables) to configure the React app, a rebuild (`git push heroku master`) is required for changes to take effect. These values get embedded in the JavaScript bundle during build, whereas Heroku apps should always [read config from environment variables](https://12factor.net/config).

Examples of issues this causes:
- when using [pipelines](https://devcenter.heroku.com/articles/pipelines), [promoting an app](https://devcenter.heroku.com/articles/pipelines#promoting) (e.g. from staging to production) will not pick-up new values for these variables (until a new build is triggered)
- when setting [config vars](https://devcenter.heroku.com/articles/config-vars), the value is not applied immediately; the app must be re-deployed to pick-up the new config value
- in general, [slugs](https://devcenter.heroku.com/articles/slug-compiler) are no longer reusable per Heroku's standard architecture
## Proposal

Perform replacement of environment variables in the bundle during the app start-up (via [`.profile.d/` script](https://devcenter.heroku.com/articles/buildpack-api#profile-d-scripts)), so that the current environment variables are used at runtime. 
## Caveats
- requires developer awareness of the difference between [compiled and runtime vars](https://github.com/mars/create-react-app-buildpack/blob/modular-runtime-env-vars/README.md#compile-time-vs-runtime)
- requires developer implementation of [runtime vars npm module](https://www.npmjs.com/package/@mars/heroku-js-runtime-env)
## Try it

Try the proposed runtime env vars strategy by using the buildpack branch:

``` bash
heroku buildpacks:set https://github.com/mars/create-react-app-buildpack.git#modular-runtime-env-vars
```

Then, [follow the new Runtime vs Compile-time docs](https://github.com/mars/create-react-app-buildpack/blob/modular-runtime-env-vars/README.md#environment-variables).

**Comment by eliperelman** [comment]
I vote yes. As the person who added the environment variables and a user of Heroku, I can tell you that this is the way I do it as well. If the environment variables change between environments, and these variables need to be embedded in the static app, then the app needs to be re-built for the environment. The other option may be to fetch the Heroku environment, and run `npm run build` against the cloned environment locally, then deploy that build.

<sup>For some reason I thought the script was `.profile` instead of `.profile.d`.</sup>

**Comment by eliperelman** [comment]
Also something I have been considering lately is the pain of getting these environment variables into a CRA app in the first place. For example, I've been using the Pixie add-on which exposes a `$FIXIE_URL` environment variable. It's a pain to have to add a profile script just to map it to `REACT_APP_PROXY = $FIXIE_URL`, etc.

**Comment by mars** [comment]
Great to hear from you @eliperelman 

`.profile.d` & `.profile` are similar in behavior, the docs linked above say:

> Scripts in `.profile.d/` should only be written by buildpacks. Customers intending to perform application specific initialization tasks at the time a dyno boots should use `.profile` scripts, which are guaranteed to run after the scripts in `profile.d/`.

I'm proposing to make this change at the buildpack level, which would **not** required changes to apps such as individual developers making a `.profile` script.

Re: pain of non-`REACT_APP_*` variables…  you can attach Heroku Add-ons with a custom prefix to solve this problem:

``` bash
heroku addons:create fixie --as REACT_APP_FIXIE
```

See: [Creating an add-on](https://devcenter.heroku.com/articles/managing-add-ons#creating-an-add-on)

**Comment by eliperelman** [comment]
Wow, great info, thanks!

**Comment by mars** [comment]
Try the proposed stateless build strategy by using the buildpack branch:

``` bash
heroku buildpacks:set https://github.com/mars/create-react-app-buildpack.git#stateless-build
```

Let me know how it goes 🍃

**Comment by mars** [comment]
The biggest question remaining is: do folks have builds that take close to or longer than 60-seconds?

**Comment by mars** [comment]
~~Just realized, we can use the new [release phase](https://devcenter.heroku.com/articles/release-phase) feature to perform the build, avoiding boot timeout problems.~~

Edit: release phase is not a solution, because [changes to the slug filesystem are not persisted](https://devcenter.heroku.com/articles/release-phase#design-considerations).

**Comment by mars** [comment]
I have a new solution which does not suffer from the same build-on-startup issue. Although, it does require adding some code to your React app.

See: [runtime-env-vars branch](https://github.com/mars/create-react-app-buildpack/tree/runtime-env-vars#runtime-configuration) for the technique.

Try it out by setting your buildpack to the branch:

``` bash
heroku buildpacks:set https://github.com/mars/create-react-app-buildpack.git#runtime-env-vars
```

**Comment by mars** [comment]
Updated with a new solution; see issue description. Planning to merge as the next version of create-react-app (0.7) is released. Feedback welcome!

**Comment by mars** [comment]
Resolved by #14
